src/collections/collection.ts

```
@@ -65,340 +65,187 @@ export class Collection {
    * @param cb
    * @returns Promise
    */
-  async insertOne(doc: Record<string, any>, options?: any, cb?: DocumentCallback) {
+  async insertOne(document: Record<string, any>, options?: any, cb?: DocumentCallback) {
     ({ options, cb } = setOptionsAndCb(options, cb));
     return executeOperation(async (): Promise<InsertOneResult> => {
-      addDefaultId(doc);
-      const { data } = await this.httpClient.put(`/${doc._id}`, doc, options);
-
-      return {
-        acknowledged: true,
-        insertedId: data.documentId
+      let command = {
+        insertOne : {
+            document
+        }
       };
+      const resp = await this.httpClient.executeCommand(command);
+      if(resp.errors && resp.errors.length > 0){
+        logger.error('Error returned from server %s', JSON.stringify(resp.errors));
+        return {
```

In Mongoose, the expectation is that `insertOne()` will throw an error if the server returned an error. So this should probably `throw`.

---

Got it. A couple of questions.
1. Is that the expectation for all of the APIs? Can we throw a custom exception (something called `AstraException`) wrapping the error message and make it consistent across all of the APIs?
2. In which case, `acknowledged` is generally set to `false` vs throwing an error?

---

1. In Mongoose 6, the expectation is that the driver layer will call the `callback()` with an error if the server threw an error. Mongoose does **not** expect the driver layer to return an object with an error description.
2. `acknowledged` is false if [write concern was disabled](https://www.mongodb.com/docs/manual/reference/method/db.collection.updateMany/#returns). In other words, if you asked the server for a "fire and forget" write where the server confirms it received the write operation, but doesn't confirm whether it persisted the write operation yet, then you'll get back `acknowledged: false`. `acknowledged: false` is not an error.

---

@vkarpov15 
- I replaced all the `acknowledged: false` return statements (when there is a failure) with 'throw new Error' now. Please review and let me know if the latest change looks good with respect to error handling.
- And now I understand better on the `acknowledged` field in the response, thanks for the detailed explanation. This is going to be 'true' in our case here for all the APIs completed so far, since there is no fire and forget kind of APIs implemented so far.